public/embed.js

```diff
@@ -19,44 +19,65 @@
   };
 
   ready(function () {
-    /** @type {Map<number, HTMLIFrameElement>} */
-    var iframes = new Map();
+    /** @type {Map<number, HTMLQuoteElement>} */
+    var embeds = new Map();
 
     window.addEventListener('message', function (e) {
       var data = e.data || {};
 
-      if (typeof data !== 'object' || data.type !== 'setHeight' || !iframes.has(data.id)) {
+      if (typeof data !== 'object' || data.type !== 'setHeight' || !embeds.has(data.id)) {
         return;
       }
 
-      var iframe = iframes.get(data.id);
+      var container = embeds.get(data.id);
 
-      if(!iframe) return;
+      if (!container) return;
 
-      if ('source' in e && iframe.contentWindow !== e.source) {
+      var iframe = container.querySelector('iframe');
+
+      if (!iframe || ('source' in e && iframe.contentWindow !== e.source)) {
         return;
       }
 
       iframe.height = data.height;
+
+      var placeholder = container.querySelector('a');
+
+      if (!placeholder) return;
+
+      container.removeChild(placeholder);
     });
 
-    document.querySelectorAll('iframe.mastodon-embed').forEach(iframe => {
-      // select unique id for each iframe
+    document.querySelectorAll('blockquote.mastodon-embed').forEach(container => {
+      // Select unique id for each iframe
       var id = 0, failCount = 0, idBuffer = new Uint32Array(1);
-      while (id === 0 || iframes.has(id)) {
+
+      while (id === 0 || embeds.has(id)) {
         id = crypto.getRandomValues(idBuffer)[0];
         failCount++;
+
         if (failCount > 100) {
           // give up and assign (easily guessable) unique number if getRandomValues is broken or no luck
-          id = -(iframes.size + 1);
+          id = -(embeds.size + 1);
           break;
         }
       }
 
-      iframes.set(id, iframe);
+      embeds.set(id, container);
+
+      var iframe = document.createElement('iframe');
+      var embedUrl = new URL(container.getAttribute('data-embed-url'));
+
+      if (embedUrl.protocol !== 'https:') return;
 
+      iframe.src = embedUrl.toString();
+      iframe.width = container.clientWidth;
+      iframe.height = 0;
       iframe.scrolling = 'no';
+      iframe.allowfullscreen = true;
+      iframe.style.border = 0;
       iframe.style.overflow = 'hidden';
+      iframe.style.display = 'block';
```

Comment:
As this stands, I'm afraid this could possibly get used by an attacker in an XSS attack. I believe this would enable an attacker to execute arbitrary javascript in any page that includes this script as long as they are able to inject HTML with `class` and `data-embed-url`. This is pretty unlikely, but still offers attackers a wider attack surface they would otherwise not have access to if the HTML they can inject is otherwise sanitized.

Using `sandbox` on the `iframe` may help, but I am not sure to which extent.

Stricter checks on the URL may also help, but it probably cannot be made useful without restricting the script to  the specific Mastodon server the embed is from, which will be impractical for several reasons.

Reply:
I have added `sandbox="allow-scripts"` to the iframe, however, at least with `localhost`, this seems to break everything. All I'm getting in the browser console is that a JS file is "not eligible for integrity checks since it's neither CORS-enabled nor same-origin".

Reply:
Thinking more about it, I think:
- `sandbox="allow-scripts allow-same-origin"` should make that work and guard against XSS when the `data-embed-url` is not of the same origin as the page including the embed
- we can pass a data attribute to the script itself with a list of allowed domains/URL prefixes, this way an attacker would have to be able to modify the `script` tag itself (thus already having XSS) to be able to use the embed script to widen their attack surface

```html
<blockquote data-embed-url="https://foo/bar" …>
<script data-allowed-prefixes="https://foo/" src="https://scripts.joinmastodon.org/embed.js" integrity=… />
```

```javascript
const allowedPrefixes = (document.currentScript && document.currentScript.tagName.toUpperCase() === 'SCRIPT') ? document.currentScript.dataset.allowedPrefixes.split(' ') : [];
```